pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java

```diff
@@ -153,20 +158,56 @@ public void tearDown() {
 
   // TODO: name the test using testCaseName for testng reports
   @Test(dataProvider = "testResourceQueryTestCaseProviderInputOnly")
-  public void testQueryTestCasesWithH2(String testCaseName, String sql)
+  public void testQueryTestCasesWithH2(String testCaseName, String sql, String expect)
       throws Exception {
     // query pinot
-    List<Object[]> resultRows = queryRunner(sql);
+    Optional<List<Object[]>> resultRows = runQuery(sql, expect);
+    if (!resultRows.isPresent()) {
+      // successfully caught error
+      return;
+    }
+
     // query H2 for data
     List<Object[]> expectedRows = queryH2(sql);
-    compareRowEquals(resultRows, expectedRows);
+    compareRowEquals(resultRows.get(), expectedRows);
   }
 
   @Test(dataProvider = "testResourceQueryTestCaseProviderBoth")
-  public void testQueryTestCasesWithOutput(String testCaseName, String sql, List<Object[]> expectedRows)
+  public void testQueryTestCasesWithOutput(String testCaseName, String sql, List<Object[]> expectedRows, String expect)
       throws Exception {
-    List<Object[]> resultRows = queryRunner(sql);
-    compareRowEquals(resultRows, expectedRows);
+    Optional<List<Object[]>> resultRows = runQuery(sql, expect);
```

Comment:
this refactoring seems convoluted. can't we simply do:
```suggestion
    List<Object[]> resultRows;
    if (except != null) {
      try {
        resultRows = queryRunner(SQL);
        Assert.fail("Expected error with message '" + except + "'. But instead rows were returned: " + ...);
      } catch (Exception e) {
        Assert.assertTrue(...PatternMatch)
      }
    } else {
      resultRows = queryRunner(sql);
    }
```

IIRC Assert.fail throws AssertionError which is not an Exception so it won't be caught in the catch Exception clause.

Reply:
> IIRC Assert.fail throws AssertionError which is not an Exception so it won't be caught in the catch Exception clause.

ah good call!

> this refactoring seems convoluted. can't we simply do:

I can, but then I need to do it twice (once for the h2 compat ones, ones for the output checkers)

Reply:
why would you need to do this for H2? if you are asserting failure. you don't need H2 right?

(IIUC this is probably b/c the output can't be some of the queries list). for now we can do both but we will refactor that out later if needed.

Reply:
I made it a bit cleaner, LMKWYT after I push 😉

Reply:
looks good